### PR TITLE
Add `depth` parameter and pipe through the event creation methods

### PR DESCRIPTION
**Pull request by MadLittleMods**
Add `depth` parameter and pipe through the event creation methods. Also add `auth_event_ids` and `prev_event_ids` in some places as well so I don't have to go through PR splitting conflict hell.

Split out from https://github.com/matrix-org/synapse/pull/9247

Call stacks:
```
create_and_send_nonmember_event
create_event
create_new_client_event
builder.build
```

```
update_membership
update_membership_locked
_local_membership_update
create_event
```

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

### CODE REVIEW

File: synapse/handlers/room_member.py
Comment (MadLittleMods) [inline synapse/handlers/room_member.py:565]: > New case is here where we override the prev_event_ids if passed in explicitly
>
> https://github.com/matrix-org/synapse/pull/9247/files#r610994535

File: synapse/handlers/room_member.py
Comment (MadLittleMods) [inline synapse/handlers/room_member.py:567]: > All of this code is just nested under the `else` conditional and not changed.
>
> https://github.com/matrix-org/synapse/pull/9247/files#r610994518

File: synapse/storage/databases/main/event_federation.py
Comment (MadLittleMods) [inline synapse/storage/databases/main/event_federation.py:674]: A bit of new logic here. We also return the event ID for the event with the greatest depth

File: synapse/events/builder.py
Comment (MadLittleMods) [inline synapse/events/builder.py:173]: The smarts of `inherit_depth` are here 🎈

File: synapse/handlers/room_member.py
Comment (erikjohnston) [inline synapse/handlers/room_member.py:727]: I wonder whether instead of indenting all this we should return early?

**Review by erikjohnston — COMMENTED** [review]
These look like changes that would work :)

File: synapse/handlers/room_member.py
Comment (MadLittleMods) [reply synapse/handlers/room_member.py]: I've changed it to return early. It's pretty hard to spot the return in the middle of the function.

I'm thinking about moving all of this nested logic to its own function so it's easier to read here.

Comment (MadLittleMods) [reply synapse/handlers/room_member.py]: Looking closer, that bunch of nested logic has its own returns. Can leave it for now ⏩

**Comment by MadLittleMods** [comment]
I feel like the remaining failing Sytest tests are just flakey. They are different when I re-run them.

Trying to get it running locally to confirm, https://github.com/matrix-org/sytest/issues/1054

File: synapse/handlers/message.py
Comment (erikjohnston) [inline synapse/handlers/message.py:803]: Can we pass these in instead? I'm wary of accidentally letting clients set it.

**Review by erikjohnston — COMMENTED** [review]
Thanks, this looks good other than tiny nit

File: synapse/handlers/message.py
Comment (MadLittleMods) [reply synapse/handlers/message.py:808]: Good idea :+1:

**Comment by MadLittleMods** [comment]
@erikjohnston Thanks for all the review so far 🎢

**Comment by erikjohnston** [comment]
I think this is good to merge, but I'm going to hold off a bit until we #9247 is a bit closer (in particular I'm wondering if we should pass in the depth explicitly, rather than have `inherit_depth`). Thanks for splitting it out from that PR though, it was helpful to be able to review it separately :)

**Comment by MadLittleMods** [comment]
@erikjohnston I've updated this PR and the [base PR](https://github.com/matrix-org/synapse/pull/9247) to pass around `depth` directly :+1:

File: synapse/events/builder.py
Comment (erikjohnston) [inline synapse/events/builder.py:142]: May as well move this into the `if depth is None`?

File: synapse/storage/databases/main/event_federation.py
Comment (erikjohnston) [reply synapse/storage/databases/main/event_federation.py:674]: Presumably this new logic will get used when we add the bulk insert API? (It's not used in this PR anyomre)

**Review by erikjohnston — APPROVED** [review]
This looks good, thanks, but again let's wait for the other PR before landing this

File: synapse/storage/databases/main/event_federation.py
Comment (MadLittleMods) [reply synapse/storage/databases/main/event_federation.py:674]: It is still used in https://github.com/matrix-org/synapse/pull/9247. I can remove here if preferred.

Comment (erikjohnston) [reply synapse/storage/databases/main/event_federation.py:674]: I think its ok to leave here.

File: tests/handlers/test_presence.py
Comment (MadLittleMods) [inline tests/handlers/test_presence.py:866]:
```suggestion
        event = self.get_success(builder.build(prev_event_ids, None, None))
```

**Pull request closed without merging**
